pkg/controller/machine_safety.go

```diff
@@ -0,0 +1,482 @@
+				if err != nil {
+					glog.Error("Error parsing time: ", err)
+					wg.Done()
+					return
+				}
+
+				//glog.Info(time.Since(timestamp))
+
+				if ms.Status.ReadyReplicas == ms.Spec.Replicas {
+					for {
+						// Get the latest version of the machineSet so that we can avoid conflicts
+						ms, err := c.controlMachineClient.MachineSets(ms.Namespace).Get(ms.Name, metav1.GetOptions{})
+						if err != nil {
+							// Some error occued while fetching object from API server
+							glog.Error(err)
+							break
+						}
+						clone := ms.DeepCopy()
+						delete(clone.Annotations, LastReplicaUpdate)
+						_, err = c.controlMachineClient.MachineSets(clone.Namespace).Update(clone)
+						if err == nil {
+							break
+						}
+						// Keep retrying until update goes through
+						glog.Warning("Updated failed, retrying")
+					}
+				} else if time.Since(timestamp) > timeout {
+					message := "MachineSet has timed out while scaling replicas"
+					c.freezeOvershootingMachineObjects(ms, TimeoutOccurred, message)
+				}
+			}
+		}
+	}
+
+	wg.Done()
+}
+
+// checkAWSMachineClass checks for orphan VMs in AWSMachinesClasses
+func (c *controller) checkAWSMachineClass(wgForCheckVM *sync.WaitGroup) {
+	AWSMachineClasses, err := c.awsMachineClassLister.List(labels.Everything())
+	if err != nil {
+		glog.Error("Safety-Net: Error getting machineClasses")
+		wgForCheckVM.Done()
+		return
+	}
+
+	for _, machineClass := range AWSMachineClasses {
+
+		var machineClassInterface interface{}
+		machineClassInterface = machineClass
+
+		c.checkMachineClass(
+			machineClassInterface,
+			machineClass.Spec.SecretRef,
+			machineClass.Name,
+			machineClass.Kind,
+		)
+	}
+
+	wgForCheckVM.Done()
+}
+
+// checkMachineClass checks a particular machineClass for orphan instances
+func (c *controller) checkMachineClass(
+	machineClass interface{},
+	secretRef *corev1.SecretReference,
+	className string,
+	classKind string) {
+
+	// Get secret
+	secret, err := c.getSecret(secretRef, className)
+	if err != nil || secret == nil {
+		glog.Errorf("Secret reference not found for MachineClass: %q", className)
+		return
+	}
+
+	// Dummy driver object being created to invoke GetVMs
+	dvr := driver.NewDriver(
+		"",
+		secret,
+		classKind,
+		machineClass,
+		"",
+	)
+	listOfVMs := dvr.GetVMs("")
+
+	// Making sure that its not a VM just being created, machine object not yet updated at API server
+	if len(listOfVMs) > 1 {
+		stopCh := make(chan struct{})
+		if !cache.WaitForCacheSync(stopCh, c.machineSynced) {
+			glog.Errorf("Timed out waiting for caches to sync")
+			return
+		}
+	}
+
+	for _, vm := range listOfVMs {
+		machine, err := c.controlMachineClient.Machines(c.namespace).Get(vm.MachineName, metav1.GetOptions{})
+		if err != nil && !apierrors.IsNotFound(err) {
+			// Any other types of errors
+			glog.Error("Safety-Net: Error getting machines ", err)
+		} else if machine.Spec.ProviderID != vm.MachineID || err != nil {
```

Comment:
I think we should test it with orphan VMs with no MachineObject backing it, or shouldn't it panic?

Reply:
Done.